proof-generator/src/lib.rs

```diff
@@ -0,0 +1,170 @@
+use anyhow::anyhow;
+use coins_bip32::path::DerivationPath;
+use ethers::{
+    prelude::Signer,
+    providers::{Http, Middleware, Provider},
+    signers::{HDPath, Ledger},
+    types::{Signature, H256},
+};
+use serde::{Deserialize, Serialize};
+use std::{
+    convert::TryFrom,
+    fmt::{Debug, Display},
+    fs,
+    io::Write,
+    path::PathBuf,
+    process::{Command, Stdio},
+    str,
+};
+
+/// The options allowed to be provided to the CLI
+#[derive(Debug, Clone)]
+pub struct Options {
+    /// GPG key to sign the proof.
+    pub gpg_key: String,
+    /// Output path of created proof
+    pub output: PathBuf,
+    /// RPC url
+    pub rpc_url: String,
+    /// Account derivation path when using a Ledger hardware wallet.
+    pub ledger_hdpath: Option<DerivationPath>,
+    /// Keystore file containing encrypted private key (default: none).
+    pub keystore: Option<PathBuf>,
+}
+
+/// The proof to be constructed and signed
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct Proof {
+    /// GPG signature of predefined proof
+    gpg_sig: String,
+    /// ETH signature of predefined proof
+    eth_sig: Signature,
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum Error {
+    /// No wallet specified.
+    #[error("no wallet specified")]
+    NoWallet,
+    /// Not able to retrieve block .
+    #[error("not able to retrieve block")]
+    NoBlock,
+    /// Not able to retrieve block hash .
+    #[error("not able to retrieve block hash")]
+    NoBlockHash,
+    /// ETH signature failed
+    #[error("eth signature failed")]
+    ETHSigFailed,
+    /// GPG signature failed
+    #[error("{0}")]
+    GPGSigFailed(String),
+}
+
+/// Sign a message with a GPG private key using the GPG CLI
+fn gpg_sign(key: &str, message: &str) -> anyhow::Result<String> {
+    let mut gpg = Command::new("gpg")
+        .arg("--clear-sign")
+        .arg("-u")
+        .arg(key)
+        .stdin(Stdio::piped())
+        .stderr(Stdio::piped())
+        .stdout(Stdio::piped())
+        .spawn()?;
+
+    gpg.stdin.as_mut().unwrap().write_all(message.as_bytes())?;
+
+    let output = gpg.wait_with_output()?;
+    if output.status.success() {
+        let raw_output = String::from_utf8(output.stdout)?;
+        Ok(raw_output)
+    } else {
+        Err(anyhow!(Error::GPGSigFailed(String::from_utf8(
+            output.stderr
+        )?)))
```

Comment:
When getting the output, I'd use `from_utf8_lossy`, as it's infallible and will just replace unprintable characters with boxes.

Reply:
You can also change it in the `else` branch